### ISSUE TITLE
ServiceAccount not garbage collected when owned by another resource

### ISSUE THREAD

**Issue by teoincontatto** [kubernetes/kubernetes#120960]
### What happened?

A ServiceAccount with `.metadata.ownerReferences` set to another resource (tested with ServiceAccount, Service, Deployment, ConfigMap and Secret) get not garbage collected when the owner resource is removed.

### What did you expect to happen?

The owned ServiceAccount to be removed when the owner resource get removed.

### How can we reproduce it (as minimally and precisely as possible)?

Here is a script to reproduce the issue:

```
#!/bin/sh

NAMESPACE=default
OWNER_KIND=Service
OWNED_KIND=ServiceAccount
CONTROLLER=true
BLOCK_OWNER_DELETION=true

get_apiversion() {
  echo "$(
    { kubectl explain "$1" | grep '^GROUP:' | tr -d ' ' | cut -d : -f 2 && printf /; } | tr -d '\n'
    )$(
    kubectl explain "$1" | grep '^VERSION:' | tr -d ' ' | cut -d : -f 2
    )"
}

get_kind_payload() {
  case "$1" in
    ('Service'):
      echo "
spec:
  ports:
  - port: 8080
"
    ;;
    ('Deployment'):
      echo "
spec:
  selector:
    matchLabels:
      test: test
  template:
    metadata:
      labels:
        test: test
    spec:
      containers:
      - name: test
        image: alpine
        command: [ sleep, '3600' ]
"
    ;;
  esac
}

OWNER_APIVERSION="$(get_apiversion "$OWNER_KIND")"
OWNED_APIVERSION="$(get_apiversion "$OWNED_KIND")"
OWNER_KIND_PAYLOAD="$(get_kind_payload "$OWNER_KIND")"
OWNED_KIND_PAYLOAD="$(get_kind_payload "$OWNED_KIND")"

kubectl delete --ignore-not-found -n "$NAMESPACE" "$OWNER_KIND" owner
kubectl delete --ignore-not-found -n "$NAMESPACE" "$OWNED_KIND" owned

echo

echo "* Creating resource $OWNER_KIND owner"

echo

cat << EOF > /tmp/owner.yaml
apiVersion: $OWNER_APIVERSION
kind: $OWNER_KIND
metadata:
  name: owner
$OWNER_KIND_PAYLOAD
EOF

kubectl create -n "$NAMESPACE" -f /tmp/owner.yaml -o yaml

echo

echo "* Creating resource $OWNED_KIND owned"

echo

cat << EOF > /tmp/owned.yaml
apiVersion: $OWNED_APIVERSION
kind: $OWNED_KIND
metadata:
  name: owned
  ownerReferences:
  - $(kubectl get -n "$NAMESPACE" "$OWNER_KIND" owner -o json | jq -c "{ apiVersion: .apiVersion, kind: .kind, name: \"owner\", uid: .metadata.uid, controller: $CONTROLLER, blockOwnerDeletion: $BLOCK_OWNER_DELETION }")
$OWNED_KIND_PAYLOAD
EOF

kubectl create -n "$NAMESPACE" -f /tmp/owned.yaml -o yaml

echo

echo "* Listing resources"

echo

kubectl get -n "$NAMESPACE" "$OWNER_KIND"

echo

kubectl get -n "$NAMESPACE" "$OWNED_KIND"

echo

echo "* Delete $OWNER_KIND owner waiting for deletion of child resources"

echo

kubectl delete -n "$NAMESPACE" "$OWNER_KIND" owner --wait

echo

echo "* Waiting for deletion of $OWNED_KIND owned"

echo

kubectl wait --for delete --timeout 10s -n "$NAMESPACE" "$OWNED_KIND" owned -o name

echo

echo "* Listing resources"

echo

kubectl get -n "$NAMESPACE" "$OWNER_KIND"

echo

kubectl get -n "$NAMESPACE" "$OWNED_KIND"

echo

echo "Cleanup"

echo

kubectl delete --ignore-not-found -n "$NAMESPACE" "$OWNED_KIND" owned
```

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: v1.28.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.27.3
```

</details>

### Cloud provider

<details>
No provider, test environment was using kind
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
AME="Ubuntu"
VERSION="20.04.6 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.6 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal

$ uname -a
Linux X1CarbonGen10 5.14.0-1059-oem #67-Ubuntu SMP Mon Mar 13 14:22:10 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
```

</details>

### Install tools

<details>
kind version
kind v0.20.0 go1.20.4 linux/amd64

kind create cluster
</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by liggitt** [comment]
/remove-sig auth
/sig api-machinery

GC is generically implemented across all API types, not serviceaccount-specific

**Comment by liggitt** [comment]
I'm not able to reproduce:

```sh
# create service account and configmap
kubectl create sa test -o yaml
kubectl create configmap test -o yaml
echo

# grab the configmap uid
uid=$(kubectl get configmap test -o jsonpath='{.metadata.uid}')
echo $uid
echo

# add an owner reference
kubectl patch sa test -p '{"metadata":{"ownerReferences":[{"apiVersion":"v1","kind":"ConfigMap","name":"test","uid":"'$uid'"}]}}' -o yaml
echo

# make sure the service account still exists
sleep 5
kubectl get sa test
echo

# delete the configmap
kubectl delete configmap test
echo

# make sure the service account gets cleaned up
sleep 5
kubectl get sa test
echo
```

output is
```
apiVersion: v1
kind: ServiceAccount
metadata:
  creationTimestamp: "2023-10-03T21:05:13Z"
  name: test
  namespace: default
  resourceVersion: "649"
  uid: d7060792-a5af-4aac-b4e5-efb6587707b8
apiVersion: v1
kind: ConfigMap
metadata:
  creationTimestamp: "2023-10-03T21:05:13Z"
  name: test
  namespace: default
  resourceVersion: "650"
  uid: cf7f3363-c371-4841-99c6-9f24ae7df781

cf7f3363-c371-4841-99c6-9f24ae7df781

apiVersion: v1
kind: ServiceAccount
metadata:
  creationTimestamp: "2023-10-03T21:05:13Z"
  name: test
  namespace: default
  ownerReferences:
  - apiVersion: v1
    kind: ConfigMap
    name: test
    uid: cf7f3363-c371-4841-99c6-9f24ae7df781
  resourceVersion: "651"
  uid: d7060792-a5af-4aac-b4e5-efb6587707b8

NAME   SECRETS   AGE
test   0         5s

configmap "test" deleted

Error from server (NotFound): serviceaccounts "test" not found
```

**Comment by liggitt** [comment]
also tried with `"controller":true,"blockOwnerDeletion":true` and couldn't reproduce

**Comment by teoincontatto** [comment]
Re-tested in multiple Kubernetes versions (still using kind 0.20.0):

* v1.20.15 - OK
* v1.21.14 - OK
* v1.22.17 - OK
* v1.23.17 - OK
* v1.24.15 - OK
* v1.25.11 - OK
* v1.26.6 - OK
* v1.27.3 - OK

Seems I can not reproduce this anymore. A bit weird since I was able to reproduce this before with multiple Kubernetes versions. Let's close this, if I found any way to reproduce it in the future will re-open. Thanks for your time to look into this @liggitt !
